### PR TITLE
Disable default and ambient attribute at runtime with a feature switch

### PR DESCRIPTION
**Pull request by LakshanF**
Disable design time attributes being used at runtime via a feature switch.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/ComponentModel/DefaultValueAttribute.cs [hunk 1]
```diff
@@ -35,6 +39,12 @@ public class DefaultValueAttribute : Attribute
             // The null check and try/catch here are because attributes should never throw exceptions.
             // We would fail to load an otherwise normal class.
 
+            if (!IsSupported)
+            {
+                Debug.Assert(!IsSupported, "Runtime instantiation of this attribute is not allowed.");
```
Comment (steveharter) [inline src/libraries/System.Private.CoreLib/src/System/ComponentModel/DefaultValueAttribute.cs:44]: Assert vs. throw approaches - was the Assert approach taken to be backwards compatible here?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AmbientValueAttribute.cs [hunk 2]
```diff
@@ -22,6 +24,13 @@ public AmbientValueAttribute([DynamicallyAccessedMembers(DynamicallyAccessedMemb
...
+
+            if (!IDesignerHost.IsSupported)
+            {
+                Debug.Assert(!IDesignerHost.IsSupported, "Runtime instantiation of this attribute is not allowed.");
+                return;
+            }
+
             try
             {
                 Value = TypeDescriptor.GetConverter(type).ConvertFromInvariantString(value);
```
Comment (steveharter) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AmbientValueAttribute.cs:36]: Was this (and the case for DefaultValueAttribute below) causing an error at runtime or is this to help with linkability (e.g. https://github.com/dotnet/runtime/issues/97842)?

File: src/libraries/System.Private.CoreLib/src/System/ComponentModel/DefaultValueAttribute.cs [hunk 3]
```diff
@@ -22,6 +23,9 @@ public class DefaultValueAttribute : Attribute
         // Delegate ad hoc created 'TypeDescriptor.ConvertFromInvariantString' reflection object cache
         private static object? s_convertFromInvariantString;
 
+        [FeatureSwitchDefinition("System.ComponentModel.DefaultValueAttribute.IsSupported")]
+        internal static bool IsSupported => AppContext.TryGetSwitch("System.ComponentModel.DefaultValueAttribute.IsSupported", out bool isSupported) ? isSupported : true;
```
Comment (steveharter) [inline src/libraries/System.Private.CoreLib/src/System/ComponentModel/DefaultValueAttribute.cs:27]: If this is made `readonly` then the JIT should be able to remove code as well (for cases when the trimmer isn't used).

@sbomer is my `readonly` comment above still applicable? Do we have guidance for this?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AmbientValueAttribute.cs [hunk 4]
```diff
@@ -22,6 +24,13 @@ public AmbientValueAttribute([DynamicallyAccessedMembers(DynamicallyAccessedMemb
         {
             // The try/catch here is because attributes should never throw exceptions. We would fail to
             // load an otherwise normal class.
+
+            if (!IDesignerHost.IsSupported)
+            {
+                Debug.Assert(!IDesignerHost.IsSupported, "Runtime instantiation of this attribute is not allowed.");
```
Comment (vitek-karas) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AmbientValueAttribute.cs:30]: Just curious - it feels weird asserting that the condition of the if is true in the true branch...

File: src/libraries/System.Private.CoreLib/src/System/ComponentModel/DefaultValueAttribute.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/System.Private.CoreLib/src/System/ComponentModel/DefaultValueAttribute.cs:27]: Hmm, in addition to perhaps supporting JIT, a `static readonly` field accessed by the property would cache the value, so `AppContext.TryGetSwitch()` doesn't get called every time (again, only for non-trimmed scenarios).

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AmbientValueAttribute.cs (hunk 4 above)
Comment (LakshanF) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AmbientValueAttribute.cs:30]: The assert is meant to give an indication of the problem of using this attribute when the feature switch is set. Throwing (with an appropriate message) here would be ideal but it seems prudent to respect the comments on the type. Consumers (like WinForms) could also generate a warning via its code generator when it detects use of this attribute at runtime if the feature switch is set.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AmbientValueAttribute.cs (hunk 2 above)
Comment (LakshanF) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AmbientValueAttribute.cs:36]: Yes, this is to help with #97842 (trimming with the corresponding feature switch is set).

File: src/libraries/System.Private.CoreLib/src/System/ComponentModel/DefaultValueAttribute.cs (hunk 1 above)
Comment (LakshanF) [reply src/libraries/System.Private.CoreLib/src/System/ComponentModel/DefaultValueAttribute.cs:44]: Yes, respecting the comments on the type so as not to throw

**Review by steveharter — APPROVED** [review]

**Pull request merged**